### ISSUE TITLE
Enable native double buffering for all non-native windows by default

### ISSUE THREAD

**Issue by vadz** [wxWidgets/wxWidgets#22846]
There doesn't seem to be any reason _not_ to enable double buffering, i.e. set `WS_EX_COMPOSITED` for all generic windows in wxMSW (some native windows don't work with this style), so we should try doing it and turning `wxALWAYS_NATIVE_DOUBLE_BUFFER` to 1 for wxMSW as well.

**Comment by vadz** [comment]
This was actually done in #22851, but see #22876 now.

**Comment by Tommykrul** [comment]
Unless I am misunderstanding how to enable proper double buffering in wxWidgets on Windows, I belief this change makes some incorrect assumptions, at least in the current implementation.

We have some internal tools that make liberal use of custom painted widgets. These controls are all implemented using SetBackgroundStyle(wxBG_STYLE_PAINT) / wxAutoBufferedPaintDC() / wxGCDC(); Starting with 3.3.0, all these controls exhibit severe flicking, very slow rendering, and seem to refresh after every individual paint operation (items appear one by one and the view takes several seconds to update). The same code linking against 3.2.0 shows no flickering and performance is fast as expected. The problem only occurs on Windows. macOS renders fine. 

Reading through the code changes, it seems like the assumption is being made that Windows now supports "native double buffering" through calling WS_EX_COMPOSITED (wxWindow::SetDoubleBuffered(true)) on the windows. And as such wxALWAYS_NATIVE_DOUBLE_BUFFER is now turned on for Windows as well. This in turn switches wxAutoBufferedPaintDC from wxBufferedPaintDC to simply wxPaintDC.

The problem as far as I can tell is that does not provide a generalized double buffering solution for windows. While I don't know the exact behavior or implementation of WS_EX_COMPOSITED, from what I can find online the general consensus seems to be that WS_EX_COMPOSITED only double-buffers the individual controls and that it seems to be a custom solve for certain native widgets. It does not seem to turn on double buffering for the entire panel.

We tried to explicitly enable SetDoubleBuffered(true) on any and all panels that do custom drawing using a wxGCDC and nothing we tried resulted in flicker free drawing. Conversely, if we hard code our wxDC to use a wxBufferedPaintDC, all flicking is gone and rendering performance is normal again. 

Long story short, I belief that on Windows, wxAutoBufferedPaintDC still needs to map to wxBufferedPaintDC as for as far as I can see Windows does not seem to provide a general double buffering solution through WS_EX_COMPOSITED that works for custom drawing using a wxDC. 

For this, wxALWAYS_NATIVE_DOUBLE_BUFFER will need to be set to 0 on Windows

**Comment by Tommykrul** [comment]
The other feedback might be from a co-worker. So probably not total coincidence :-) I only tried bringing in 3.3.0 recently because of an interest in trying out dark mode. In the past we were ever only on stable releases. 

I am intimately familiar with our codebase, and unless there are things that are *implicitly* turning off WS_EX_COMPOSITED, we're doing everything that I am aware of that we should be doing to turn it on. The setup isn't too complex. The relevant panel just does a SetDoubleBuffered(true), and nowhere is this turned off again. Yet this does not make the wxGCDC do any form of double buffering. It's not obvious when there is a very small number of draw operations, as the code is fast enough for the flickering to be noticeable. But once there is a large enough number of draw operations, the flickering is very obvious and severe. 

Is there anything else that one should be doing beyond SetDoubleBuffered(true) on the panel that does the drawing?

I will check today whether there is a way for me to check the double buffering flag inside of the draw function. But if it's not on, there must be something that is disabling it inside of wxWidgets that seems outside of our control.

**Comment by vadz** [comment]
Yes, wx does turn it off implicitly to work around things which don't work with it, see [your coworker's comment](https://github.com/wxWidgets/wxWidgets/commit/fcc7430610ca202b744df75dc3d7e471a1e8b8a4#commitcomment-115932505) which turned into the other issue.

**Comment by Tommykrul** [comment]
Thanks for the quick response. As per your suggestion, I checked whether double buffering is actually turned on during the wxPaintEvent. It indeed seems to be that wxWidgets is turning it off somewhere. But the code structure in our case is quite straight forward and doesn't do anything controversial that would make it obvious why it's not enabled.

class CustomPane : public wxPanel
   In the constructor:
      SetDoubleBuffered(true);
      SetBackgroundStyle(wxBR_STYLE_PAINT);
      assert(IsDoubleBuffered()); ///< this does not trigger, so at this point double buffering seems enabled

void CustomPane::onPaint(wxPaintEvent& e)
   assert(IsDoubleBuffered()) ///< This triggers. Double buffering is no longer enabled here.
   wxAutoBufferedPaintDC (paintDC(this));
   wxGCDC dc(paintDC);
   ... use dc.Draw...;

Basically, double buffering is enabled in the constructor. But once we hit our first paint event, it's disabled. No other code in the tool calls SetDoubleBuffered(). Nor does the panel have any native child widgets that could be expected to be disabling double buffering. It only does custom drawing using a GCDC on the whole panel (no child ListViews, etc). 

The one element that I could imagine being a wildcard is that the panel is managed by wxAuiManager.

Finally, if I call SetDoubleBuffered(true) inside of CustomPane::onPaint(wxPaintEvent& e), double buffering DOES seem to work.

**Comment by vadz** [comment]
Please try putting a breakpoint at `wxWindow::MSWDisableComposited()` and check if it's called from somewhere.

If it isn't, I'd really like to have a small example allowing to reproduce the problem, ideally as a patch to the aui sample. TIA!

**Comment by Tommykrul** [comment]
Thanks for the suggestion. 

I will look at this more tomorrow, but a quick test shows that wxWindow::MSWDisableComposited() is indeed invoked from wxGLCanvas. But what's important to note is for our case is that this OpenGL panel is a completely separate dockable wxPanel, though managed by the same wxAuiManager. Could it be that the issue is that MSWDisableComposited() recursively disables double buffering for all parent panels, which when docking on the same wxFrame causes double buffering to be disable for all panels?

I've attached a screenshot to show the setup. The code base is very large. Creating a minimal test case will take more time. 
![Capture](https://github.com/wxWidgets/wxWidgets/assets/7014452/b3b9ad31-4b4c-4ff8-8906-1839dce7faf5)

**Comment by Tommykrul** [comment]
After some more searching online I came across your own post from Jan 7th where you already mentioned the behavior that we seem to be running in to. As far as I can tell, this is exactly what's happening here. All panels are siblings of a wxGLCanvas and thus all panels show flickering because double buffering is being turned off. 

While unfortunate, this in and of itself wouldn't be a problem if the behavior of wxAutoBufferedPaintDC wouldn't also have changed. But with the current setup it seems quite unpredictable for users when double buffering is and isn't expected to work. It definitely had us stumped for a few days, digging through change logs and wx source to see what's going on. 

<img width="1385" alt="Screenshot 2023-06-01 at 9 02 33 PM" src="https://github.com/wxWidgets/wxWidgets/assets/7014452/aa92aec3-8c3b-49c0-8041-aee3023575f6">

**Comment by vadz** [comment]
Yes, unfortunately `WS_EX_COMPOSITED` only matters for the top level window and can't be disabled for just one child -- it has to be disabled for the entire window.

Of course, the best thing would be to find some way to make `wxGLCanvas` work with compositing, but I have no idea how to do it, unfortunately, and I'm not sure if it's even possible (can OpenGL be used to draw on the off screen buffer?). And failing that we'll have to do what is discussed in #23585 i.e. re-enable double buffering for wxMSW.

**Comment by vadz** [comment]
I've found out about `PFD_SUPPORT_COMPOSITION` today, I wonder if using it could do what we need? If you can test if enabling it allows us to remove `MSWDisableComposited()` call from `wxGLCanvas`, it would be great (to test why this call is needed, just try removing it now).

**Comment by Tommykrul** [comment]
We can try that. Can you point me to any context as to why MSWDisableComposited() was needed for wxGLCanvas at all -- aka is breaking when MSWDisableComposited is not called)? 

But also, like like you mentioned, even if we find a work around for wxGLCanvas, not being able to have a ListView as sibling widget anywhere in the UI, is a pretty major limitation that makes me think that if there is no way to get rid of ALL cases where MSWDisableComposited() needs to be invoked, there is just no way to reliably use double buffering on Windows, and in that case the only recourse may be to just have Windows use wxBufferedPaintDC again as a work around.

**Comment by dgud** [comment]
Can't you solve this by re-organizing the window layout, for me it is strange to use a GLCanvas as a parent to another window,
in Wings I always have GLCanvas as "leaf" window.  I'm kind of surprised that it works at all.

Maybe I use it as a parent for menu's and dialogs but that's it, or is that what you are doing?
(I have not tested 3.3 branch yet).

**Comment by vadz** [comment]
> We can try that. Can you point me to any context as to why MSWDisableComposited() was needed for wxGLCanvas at all -- aka what is breaking when MSWDisableComposited is not called)?

It should be pretty noticeable, AFAIR nothing was getting drawn at all without it, i.e. the window just remained black. Maybe I misremember, but there were definitely some very obvious problem with (re?)painting without it.

> But also, like you mentioned, even if we find a work around for wxGLCanvas, not being able to have a ListView as a sibling widget anywhere in the UI is a pretty major limitation

You say this, I say it's just another reason to replace it with `wxDataViewCtrl` :-/

**Comment by Tommykrul** [comment]
vadz:
> You say this, I say it's just another reason to replace it with wxDataViewCtrl :-/

True I guess. And I do think we replaced all (most?) list views with wxDataViewCtrl. But this is the problem with any software project that has a large installed user base that may be relying on legacy functionality and may not have the bandwidth to refactor parts of it. I remember refactoring our outliner control, and transitioning to a data model was a non trivial amount of work when trying to keep functional parity with the original functionality and also keep 100% consistent behavior between macOS and Windows. And while I don't disagree that wxDataViewCtrl is the better choice, I think it would need to be properly deprecated if it is known to break double buffering functionality that other parts of wxWidgets rely on in order to function. 

dgud:
> Can't you solve this by re-organizing the window layout, for me it is strange to use a GLCanvas as a parent to another window

The problem is that the GL view IS a leaf node. The parent window is a wxFrame, and all panels you see are siblings hosted by wxAuxManager. But ANY panel that contains a widget that needs to invoke MSWDisableComposited() in order to work, will disable double buffering recursively up to the root, meaning that eventually the parent wxFrame has double buffering disabled. In other words, Windows double buffering gets disabled for every panel if you have a GL panel anywhere in the UI -- except if it's a wholly separate wxFrame.

**Comment by vadz** [comment]
@Tommykrul I've actually stumbled upon a way to make `wxListCtrl` work with composited windows, see #23608. So if we can make `wxGLCanvas` work with it too, we should be able to always use `WS_EX_COMPOSITED`.

**Comment by Tommykrul** [comment]
That's fantastic to hear! Haven't had a chance yet to check wxGLCanvas / PFD_SUPPORT_COMPOSITION

**Comment by dgud** [comment]
@Tommykrul Thanks for explaining, didn't understand it reset "upward".

**Comment by MMFuba** [comment]
Hi,
i'm still being bitten by issue #15397 which has been moved here.  I use wxWidgets 3.2.4 from Msys2 on Win 10 64 Bit. I don't feel confident enough (just starting part time on c++/wxWidgets) to look into this issue myself (yet?) but I understand that this seems to be rather tricky to figure out.

So as I can't help with coding mayby I can contribute with testing... I seem to be able to reproduce this issue reliably.
Can you maybe suggest some kind of workaround for this issue? Is there a good alternative to wxGraphicsContext::GetSize() I could use for now?

**Comment by vadz** [comment]
I don't think the other issue "was moved here". If you use `wxBufferedPaintDC` explicitly, the other issue is still relevant, it's just it's low priority because

1. You don't have any real reason to use it explicitly any longer.
2. You can always use `wxWindow::GetSize()` instead.

**Comment by MMFuba** [comment]
Re 2: For now I am using wxWindow::GetSize() and it seems to solve my window resizing issues without any noticeable side effects. Thanks a lot for this nice advice!

Re 1: I am confused about you saying I wouldn't have a reason to (explicitly) use it any longer because:
- I am sure I need double buffering
- the recommended way of enabling double buffering seems to be using [wxAutoBufferedPaintDC](https://docs.wxwidgets.org/3.2/classwx_auto_buffered_paint_d_c.html)
- however [wxAutoBufferedPaintDC](https://docs.wxwidgets.org/3.2/classwx_auto_buffered_paint_d_c.html) says: [wxAutoBufferedPaintDC](https://docs.wxwidgets.org/3.2/classwx_auto_buffered_paint_d_c.html) is simply a typedef of [wxPaintDC](https://docs.wxwidgets.org/3.2/classwx_paint_d_c.html) on platforms that have native double-buffering, otherwise, it is a typedef of [wxBufferedPaintDC](https://docs.wxwidgets.org/3.2/classwx_buffered_paint_d_c.html).

So by doing the recommended thing I end up using wxBufferedPaintDC on wxMSW (Or don't I?). So how do I not have a reason to use it?

**Comment by vadz** [comment]
> Re 1: I am confused about you saying I wouldn't have a reason to (explicitly) use it any longer because:

You're commenting in an issue only relevant in master where double buffering is enabled by default in wxMSW, so you don't need to do anything at all, and in particular use `wxBufferedPaintDC`, in order to use it.
